src/codegen/compile.ml

```diff
@@ -1714,6 +1731,31 @@ module Tagged = struct
   let _branch_typed_with env ty retty branches =
     branch_with env retty (List.filter (fun (tag,c) -> can_have_tag ty tag) branches)
 
+  let allocation_barrier env get_object = 
+    (if !Flags.gc_strategy = Flags.Incremental then
+      (* performance gain by first checking the GC state *)
```

Comment:
Is that faster than just checking the state within rts_allocation_barrier? Maybe it is in our weird cost model.

Reply:
Yes, unfortunately, a relatively high difference (but need to remeasure to get the detailed number). I observed that the RTS-calls are much more expensive if the `Memory` trait is used (no extra cost if it is declared but not used). Also `not inlining` functions at the right places shows noticeable differences. I hope it is improving with our future cost model.

Reply:
We have a new cost model coming?

Reply:
I think so, the runtime system team is working on this. But not sure about the details and timeline...

Reply:
I remeasured this optimization (to have specific results): Together with the write barrier, it is a **>10%** runtime performance gain (total instruction count, average across benchmark cases). For the allocation barrier, it gives a **4%** benefit, and for the write barrier >6%.